graphene/src/document.rs

```diff
@@ -154,6 +166,27 @@ impl Document {
 		Ok(())
 	}
 
+	/// Visit each layer recursively, applies modify_shape to each non-overlay Shape
+	/// Currently used to swap between viewmodes
+	pub fn visit_all_shapes<F: FnMut(&mut Shape) -> ()>(layer: &mut Layer, modify_shape: &mut F) -> bool {
```

Comment:
It would be nice if there was a function that also cleans the upstream caches in case you want to work on anything but the root